include/swift/AST/Expr.h

```diff
@@ -553,6 +623,81 @@ class alignas(8) Expr {
   }
 };
 
+/// Helper class to capture trailing call argument labels and related
+/// information, for expression nodes that involve argument labels, trailing
+/// closures, etc.
+template<typename Derived>
+class TrailingCallArguments
+    : private llvm::TrailingObjects<Derived, Identifier, SourceLoc> {
+  using TrailingObjects = llvm::TrailingObjects<Derived, Identifier, SourceLoc>;
+  friend TrailingObjects;
+
+  Derived &asDerived() {
+    return *static_cast<Derived *>(this);
+  }
+
+  const Derived &asDerived() const {
+    return *static_cast<const Derived *>(this);
+  }
+
+  size_t numTrailingObjects(
+      typename TrailingObjects::template OverloadToken<Identifier>) const {
+    return asDerived().getNumArguments();
+  }
+
+  size_t numTrailingObjects(
+      typename TrailingObjects::template OverloadToken<SourceLoc>) const {
```

Comment:
It's correct C++, but you might need to break it up for MSVC to handle, e.g., by putting it into a separate typedef:

    typedef typename TrailingObjects::template OverloadToken<SourceLoc> SourceLocOverloadToken;

or dropping the "template" keyword just on MSVC using preprocessor macros.